File: sixthSeminar/src/main/java/org/sopt/sixthSeminar/external/S3Service.java
package org.sopt.sixthSeminar.external;

import org.sopt.sixthSeminar.config.AwsConfig;
import org.sopt.sixthSeminar.exception.ErrorMessage;
import org.sopt.sixthSeminar.exception.model.FileBadRequestException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
import software.amazon.awssdk.services.s3.model.GetUrlRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

@Component
public class S3Service {

    private final String bucketName;
    private final AwsConfig awsConfig;
    private static final List<String> IMAGE_EXTENSIONS = Arrays.asList("image/jpeg", "image/png", "image/jpg", "image/webp");
    private static final Long MAX_FILE_SIZE = 3 * 1024 * 1024L;


    public S3Service(@Value("${aws-property.s3-bucket-name}") final String bucketName, AwsConfig awsConfig) {
        this.bucketName = bucketName;
        this.awsConfig = awsConfig;
    }


    public String uploadImage(String directoryPath, MultipartFile image) throws IOException {
        final String key = directoryPath + generateImageFileName();
        final S3Client s3Client = awsConfig.getS3Client();
        validateExtension(image);
        validateFileSize(image);
        PutObjectRequest request = PutObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .build();

        RequestBody requestBody = RequestBody.fromBytes(image.getBytes());
        s3Client.putObject(request, requestBody);

        return s3Client.utilities().getUrl(GetUrlRequest.builder().bucket(bucketName).key(key).build()).toString();
    }

    public void deleteImage(String key) throws IOException {
        final S3Client s3Client = awsConfig.getS3Client();

        s3Client.deleteObject((DeleteObjectRequest.Builder builder) ->
                builder.bucket(bucketName)
                        .key(key)
                        .build()
        );
    }

    private void validateExtension(MultipartFile image) {
        String contentType = image.getContentType();
        if (!IMAGE_EXTENSIONS.contains(contentType)) {
            throw new FileBadRequestException(ErrorMessage.FILE_EXTENSION_BAD_REQUEST_EXCEPTION);
        }
    }

    private void validateFileSize(MultipartFile image) {
        if (image.getSize() > MAX_FILE_SIZE) {
            throw new FileBadRequestException(ErrorMessage.FILE_SIZE_BAD_REQUEST_EXCEPTION);
        }
    }

    private String generateImageFileName() {
        return UUID.randomUUID().toString() + ".jpg";
    }

}

File: sixthKotlin/src/main/kotlin/org/sopt/sixthKotlin/external/S3Service.kt
package org.sopt.sixthKotlin.external

import org.sopt.sixthKotlin.config.AwsConfig
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.multipart.MultipartFile
import software.amazon.awssdk.core.sync.RequestBody
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest
import software.amazon.awssdk.services.s3.model.GetUrlRequest
import software.amazon.awssdk.services.s3.model.PutObjectRequest
import java.io.IOException

import java.util.UUID

@Component
class S3Service(
    @param:Value("\${aws-property.s3-bucket-name}") private val bucketName: String,
    private val awsConfig: AwsConfig
) {
    @Throws(IOException::class)
    fun uploadImage(directoryPath: String, image: MultipartFile): String {
        val key = directoryPath + generateImageFileName()
        val s3Client = awsConfig.getS3Client()
        validateExtension(image)
        validateFileSize(image)
        val request = PutObjectRequest.builder()
            .bucket(bucketName)
            .key(key)
            .build()
        val requestBody = RequestBody.fromBytes(image.bytes)
        s3Client.putObject(request, requestBody)
        return s3Client.utilities().getUrl(GetUrlRequest.builder().bucket(bucketName).key(key).build()).toString()
    }

    @Throws(IOException::class)
    fun deleteImage(key: String?) {
        val s3Client = awsConfig.getS3Client()
        s3Client.deleteObject { builder: DeleteObjectRequest.Builder ->
            builder.bucket(
                bucketName
            )
                .key(key)
                .build()
        }
    }

    private fun validateExtension(image: MultipartFile) {
        val contentType = image.contentType
        if (!IMAGE_EXTENSIONS.contains(contentType)) {
            throw RuntimeException("이미지 확장자는 jpg, png, webp만 가능합니다.")
        }
    }

    private fun validateFileSize(image: MultipartFile) {
        if (image.size > MAX_FILE_SIZE) {
            throw RuntimeException("이미지 사이즈는 5MB를 넘을 수 없습니다.")
        }
    }

    private fun generateImageFileName(): String {
        return UUID.randomUUID().toString() + ".jpg"
    }

    companion object {
        private val IMAGE_EXTENSIONS: List<String> =
            arrayListOf("image/jpeg", "image/png", "image/jpg", "image/webp")
        private const val MAX_FILE_SIZE = 3 * 1024 * 1024L
    }
}